### PR TITLE
feat(token-selector): show default tokens for connected wallets with no balance

### PR DESCRIPTION
**Pull request by jjramirezn**
This PR adds two things:

- In the request pay flow, we indicate that wallet should be connected
- In all flows, if there is a connected wallet that has no balances, show top 6 tokens when the token selector is opened

### CODE REVIEW

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 1]
```diff
@@ -94,8 +95,23 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
...
+                : peanutTokenDetails[0].tokens
+            tokens = tokens || peanutTokenDetails[0].tokens
+            balancesToDisplay = tokens.slice(0, 6).map((token: IToken) => ({
+                ...token,
+                chainId: selectedChainID || token.chainId,
+                price: 0,
+                amount: 0,
+                currency: '',
+                value: '',
+            }))
```
Comment (Hugo0) [inline src/components/Global/TokenSelector/TokenSelector.tsx]: nit: I think ideally we'd have a comment here in the code explaining why top 6 is the choice (and not top 200 or why we dont need to rank them)

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 2]
```diff
@@ -94,8 +95,23 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
...
+        }
+
         return balancesToDisplay
-    }, [balances, safeInfo])
+    }, [balances, safeInfo, selectedChainID])
```
Comment (Hugo0) [inline src/components/Global/TokenSelector/TokenSelector.tsx]: ncie

**Review by Hugo0 — APPROVED** [review]
nicee

**Review by Hugo0 — CHANGES_REQUESTED** [review]
Actually, reviewing the behaviour in the [vercel deployment](https://peanut-ui-git-feat-selector-view-disconnected-squirrellabs.vercel.app/request/create), 2 issues:

1. when initially loading the page and clicking on tokenSelector quickly, it goes to the "explore & buy more tokens" view, and only after waiting a second or two it goes to the first page

2. I think it shouldn't go to the first page, i think these 6 tokens should be on the search page (not the initial page):
image
it's more natural for the user to view that view, and if they want a different token then they can enter some input in the search bar

Like this:
image

File: src/components/Global/TokenSelector/TokenSelector.tsx (hunk 1 above)
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: changed quite a bit, added comment on filters

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 3]
```diff
@@ -94,8 +205,23 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
...
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: you are right and I added it, thanks

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 4]
```diff
@@ -18,90 +19,203 @@ import { useWalletType } from '@/hooks/useWalletType'
...
+                                                className="absolute left-0 top-0 h-6 w-6"
+                                                fill="#999"
+                                            />
+                                        ) : (
+                                            <img
+                                                src={balance.logoURI}
+                                                className="absolute left-0 top-0 h-6 w-6"
+                                                alt="logo"
+                                                onError={(e) => {
+                                                    console.log(e)
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: i did not introduce this log, but I agree lets remove it

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 5]
```diff
@@ -18,90 +19,203 @@ import { useWalletType } from '@/hooks/useWalletType'
...
+                            <td className=" py-2">
+                                <div className="h-6 w-24 animate-colorPulse rounded-full bg-slate-700" />
+                            </td>
+                        </tr>
+                    ))
+                ) : (
+                    balances.map((balance, idx) => (
+                        <tr
+                            key={idx}
+                            className={`h-14 cursor-pointer gap-0 transition-colors hover:bg-n-3/10 ${utils.areTokenAddressesEqual(balance.address, selectedTokenAddress) && balance.chainId === selectedChainID && `bg-n-3/10`}`}
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx:58]: what?

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 6]
```diff
@@ -18,90 +19,203 @@ import { useWalletType } from '@/hooks/useWalletType'
...
-    function setToken(address: string): void {
-        setSelectedTokenAddress(address)
+    function setToken(balance: IUserBalance): void {
+        setSelectedChainID(balance.chainId)
+        setSelectedTokenAddress(balance.address)
         setTimeout(() => {
             setFilterValue('')
-            setShowFallback(false)
         }, 500)
         setVisible(false)
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: changed to 200ms to be in line with the fade-out of the modal

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 7]
```diff
@@ -112,20 +226,22 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
...
+            setSelectedBalance(
+                _balancesToDisplay.find(
+                    (balance) =>
+                        utils.areTokenAddressesEqual(balance.address, selectedTokenAddress) &&
+                        balance.chainId === selectedChainID
+                )
+            )
         }
-    }, [displayedToken])
+    }, [_balancesToDisplay, selectedTokenAddress, selectedChainID])
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: applied

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 8]
```diff
@@ -18,92 +19,204 @@ import { useWalletType } from '@/hooks/useWalletType'
...
+        setSelectedChainID(balance.chainId)
+        setSelectedTokenAddress(balance.address)
+        setVisible(false)
         setTimeout(() => {
             setFilterValue('')
-            setShowFallback(false)
-        }, 500)
-        setVisible(false)
+        }, 200) // the modal takes 200ms to close (duration-200 on leave)
     }
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: it's ok

**Comment by jjramirezn** [comment]
Big refactor to the token selector.

Merged the balance and the token search views, now the balances show with the search bar and are searchable, together with the supported tokens that have no balance. Graceful fallback for users with no tokens (only show the supported tokens from the sdk)

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 9]
```diff
@@ -18,92 +19,204 @@ import { useWalletType } from '@/hooks/useWalletType'
...
+            (!hasUserChangedChain || selectedChainID === chainId) &&
+            // if the user has typed something, only show tokens that match
+            (filterValue.length === 0 ||
+                name.toLowerCase().includes(filterValue.toLowerCase()) ||
+                symbol.toLowerCase().includes(filterValue.toLowerCase()))
 
-        return balancesToDisplay
-    }, [balances, safeInfo])
+        return _balancesToDisplay.filter(byChainAndText)
+    }, [_balancesToDisplay, filterValue, selectedChainID, hasUserChangedChain])
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx:211]: not an expensive calculation to merit useCallback

**Comment by Hugo0** [comment]
This is so much better

Just behaviour wise, can we add a "all chains" drop down at the top of the chains?

Because else we get this behaviour:
image

And also, you cannot ever go back to that (very nice) view!

File: src/components/Global/TokenSelector/Components/AdvancedButton.tsx [hunk 10]
```diff
@@ -142,23 +118,20 @@ export const AdvancedTokenSelectorButton = ({
...
                         </button>
-                    }
-                    {'xchain' === type &&
+                    )}
+                    {'xchain' === type && (
                         <button
                             aria-label="Reset token selection"
                             className="block"
                             onClick={(e) => {
-                                e.stopPropagation()  //don't open modal
```
Comment (Hugo0) [inline src/components/Global/TokenSelector/Components/AdvancedButton.tsx:161]: praise: lol, nice. consistent formatting ftw

**Comment by jjramirezn** [comment]
We will create a backlog item to show all chains and support having no chain selected, right now all views that use this component (and its associated context) expect to have a chain selected at all times (the context is initialized with a default value, right now optimism). 

Right now added the behavior to clean all filters (also the chain one) when the modal is closes

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 11]
```diff
@@ -18,92 +19,205 @@ import { useWalletType } from '@/hooks/useWalletType'
...
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx:211]: non expensive function

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 12]
```diff
@@ -18,92 +19,205 @@ import { useWalletType } from '@/hooks/useWalletType'
...
     const [visible, setVisible] = useState(false)
     const [filterValue, setFilterValue] = useState('')
+    const [selectedBalance, setSelectedBalance] = useState<IUserBalance | undefined>(undefined)
+    const [hasUserChangedChain, setUserChangedChain] = useState(false)
     const focusButtonRef = useRef<HTMLButtonElement>(null)
-    /**
-     * If the user is not connected, we show a fallback screen to search for tokens directly,
-     * instead of showing the balances.
-     */
-    const [showFallback, setShowFallback] = useState(!shouldBeConnected)
```
Comment (Hugo0) [inline src/components/Global/TokenSelector/TokenSelector.tsx:29]: praise: love the new behaviour

File: src/components/Global/TokenSelector/TokenSelector.tsx [hunk 13]
```diff
@@ -134,33 +252,28 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
                     setVisible(!visible)
                 }}
                 isVisible={visible}
-                tokenLogoUri={(IconPlaceholderChecker(selectedChainID) as string) ?? displayedToken?.logoURI}
-                tokenSymbol={displayedToken?.symbol ?? ''}
-                tokenBalance={displayedTokenBalance?.amount ?? 0}
-                chainIconUri={(IconPlaceholderChecker(selectedChainID) as string) ?? displayedChain?.icon.url}
+                tokenLogoUri={selectedBalance?.logoURI ?? ''}
```
Comment (Hugo0) [inline src/components/Global/TokenSelector/TokenSelector.tsx]: question: what if the logoUri is empty? Then we get a broken looking icon

Can we use the fallback icon we have alrdy, like in this example:
image

**Review by Hugo0 — APPROVED** [review]
Generally looks good, didn't go to deep in the review - lfg

File: src/components/Global/TokenSelector/TokenSelector.tsx (hunk 13 above)
Comment (jjramirezn) [reply src/components/Global/TokenSelector/TokenSelector.tsx]: good idea, added

**Pull request merged**